taichi/platform/metal/metal_runtime.cpp

```diff
@@ -408,6 +492,83 @@ class MetalRuntime::Impl {
   }
 
  private:
+  void init_runtime(int root_id) {
+    using namespace shaders;
+    char *addr = reinterpret_cast<char *>(runtime_mem_->ptr());
+    const int max_snodes = compiled_snodes_.max_snodes;
+    const auto &snode_descriptors = compiled_snodes_.snode_descriptors;
+    // init snode_metas
+    for (int i = 0; i < max_snodes; ++i) {
+      auto iter = snode_descriptors.find(i);
+      if (iter == snode_descriptors.end()) {
+        continue;
+      }
+      const SNodeDescriptor &sn_meta = iter->second;
+      SNodeMeta *rtm_meta = reinterpret_cast<SNodeMeta *>(addr) + i;
+      rtm_meta->element_stride = sn_meta.element_stride;
+      rtm_meta->num_slots = sn_meta.num_slots;
+      rtm_meta->mem_offset_in_parent = sn_meta.mem_offset_in_parent;
+      switch (sn_meta.snode->type) {
+        case SNodeType::root:
+          rtm_meta->type = SNodeMeta::Root;
+          break;
+        case SNodeType::dense:
+          if (sn_meta.snode->_bitmasked) {
+            rtm_meta->type = SNodeMeta::DenseBitmask;
+          } else {
+            rtm_meta->type = SNodeMeta::Dense;
+          }
+          break;
+        default:
+          TI_ERROR("Unsupported SNode type={}",
+                   snode_type_name(sn_meta.snode->type));
+          break;
+      }
+    }
+    addr += sizeof(SNodeMeta) * max_snodes;
+    // init snode_extractors
+    for (int i = 0; i < max_snodes; ++i) {
+      auto iter = snode_descriptors.find(i);
+      if (iter == snode_descriptors.end()) {
+        continue;
+      }
+      const auto *sn = iter->second.snode;
+      SNodeExtractors *rtm_ext = reinterpret_cast<SNodeExtractors *>(addr) + i;
+      for (int j = 0; j < taichi_max_num_indices; ++j) {
+        const auto &ext = sn->extractors[j];
+        rtm_ext->extractors[j].start = ext.start;
+        rtm_ext->extractors[j].num_bits = ext.num_bits;
+        rtm_ext->extractors[j].acc_offset = ext.acc_offset;
+        rtm_ext->extractors[j].num_elements = ext.num_elements;
+      }
+    }
+    addr += sizeof(SNodeExtractors) * max_snodes;
+    // init snode_lists
+    ListManager *const rtm_list_head = reinterpret_cast<ListManager *>(addr);
+    int list_data_mem_begin = 0;
+    for (int i = 0; i < max_snodes; ++i) {
+      auto iter = snode_descriptors.find(i);
+      if (iter == snode_descriptors.end()) {
+        continue;
+      }
+      const SNodeDescriptor &sn_meta = iter->second;
+      ListManager *rtm_list = reinterpret_cast<ListManager *>(addr) + i;
+      rtm_list->element_stride = sizeof(ListgenElement);
+      rtm_list->max_num_elems = sn_meta.total_num_elems_from_root;
```

Comment:
This should work fine for `bitmasked` but for sparser structures such as `ptr` and `hash`, `sn_meta.total_num_elems_from_root` might be really large. And then we need a dynamic list / memory allocator. Do you think that is doable on Metal? A basic allocator would only need
 - Preallocation of a huge chunk of GPU memory
 - A spinning lock, for e.g. the stack top of the allocator

Now I understand why starting from `bitmasked` is a good idea :-)

Reply:
That's true. I didn't expect to handle memory allocation from Metal in the beginning, and hence the design...

We might have such an allocator with preallocated GPU memory + atomic int to store the head. I need some testings here. One thing I don't really know is how to best handle memory allocation failure on the device side... There's no crash or anything to convey that error to the host side, at least not without synchronization?

Reply:
> One thing I don't really know is how to best handle memory allocation failure on the device side... There's no crash or anything to convey that error to the host side, at least not without synchronization?

For CUDA without unified memory, I'll just call an `__assertfail` intrinsic function to abort the GPU threads. I guess there might be something similar on Metal. And you are right, as long as there's no synchronization, convey the error to host is difficult. I think at least we should inform the user of OOM in debug mode when a sync is done after every kernel.

Reply:
Added a comment..